docs/api/scripting.md

```diff
@@ -1 +1,75 @@
-> Help wanted! Please [contribute to this documentation](https://github.com/jacebrowning/doorstop/edit/master/docs/api/scripting.md).
+<h1>Scripting Interface</h1>
+
+Being written in Python, Doorstop allows you to leverage the full power of Python to write scripts to manipulate requirements, run custom queries across all documents, and even inject your own validation rules.
+
+# REPL
+
+For ad hoc introspection, let Doorstop build your tree of documents in your preferred Python [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop) or notebook session:
+
+```python
+>>> import doorstop
+>>> tree = doorstop.build()
+>>> tree
+<Tree REQ <- [ TUT <- [ HLT ], LLT ]>
+>>> len(tree.documents)
+4
+>>> document = tree.find_document('REQ')
+>>> document
+Document('/Users/Browning/Documents/doorstop/reqs')
+>>> sum(1 for item in document if item.active)
+18
+```
+
+# Generic Scripting
+
+For reusable workflows, create a Python script that acts on your tree of documents:
+
+```python
+#!/usr/bin/env python
+
+import doorstop
+
+tree = doorstop.build()
+document = tree.find_document('REQ')
+count = sum(1 for item in document if item.active)
+
+print(f"{count} active items in {document}")
+```
+
+# Validation Hooks
+
+To extend the default set of [validations](../cli/validation.md) that can be performed, Doorstop provides a "hook" mechanism to simplify scripts that need to operate on multiple documents or items.
+
+For this use case, create a script to call in place of the default command-line interface:
+
+```python
+#!/usr/bin/env python
+
+
+import sys
+from doorstop import build, DoorstopInfo, DoorstopWarning, DoorstopError
+
+
+def main():
+    tree = build()
+    success = tree.validate(document_hook=check_document, item_hook=check_item)
```

Comment:
I'll open an issue after this is merged, but the idea of a "plugin" system is that these callbacks could be passed automatically if a certain convention is met (e.g. `plugin.py` exists which defines `check_document` or `check_item`).

Reply:
For a plugin system it would be nice to be able to register custom sub-commands to doorstop with custom sub-command arguments.

I think you need also a generic tree iterator which doesn't perform the default validations.

Reply:
Yeah, I could see that being useful. I'm going to merge this as-is, which documents the existing functionality.